src/Query/Builder.php

```diff
@@ -61,13 +62,23 @@ public function update(array $values)
     /**
      * @inheritDoc
      */
-    public function updateOrInsert(array $attributes, array $values = [])
+    public function updateOrInsert(array $attributes, array|callable $values = [])
     {
-        if (! $this->where($attributes)->exists()) {
+        $exists = $this->where($attributes)->exists();
+
+        if ($values instanceof Closure) {
+            $values = $values($exists);
+        }
```

Comment:
Just doing the same thing as parent.

https://github.com/laravel/framework/blob/11.x/src/Illuminate/Database/Query/Builder.php#L3755